python/deltalake/table.py

```diff
@@ -525,6 +526,62 @@ def optimize(
     ) -> "TableOptimizer":
         return TableOptimizer(self)
 
+    def merge(
+        self,
+        source: Union[pyarrow.Table, pyarrow.RecordBatch, pyarrow.RecordBatchReader],
+        predicate: str,
+        source_alias: Optional[str] = None,
+        target_alias: Optional[str] = None,
+        error_on_type_mismatch: bool = True,
+    ) -> "TableMerger":
+        """Pass the source data which you want to merge on the target delta table, providing a
+        predicate in SQL query like format. You can also specify on what to do when the underlying data types do not
+        match the underlying table.
+
+        Args:
+            source (pyarrow.Table | pyarrow.RecordBatch): source data
+            predicate (str): SQL like predicate on how to merge
+            source_alias (str): Alias for the source table
+            target_alias (str): Alias for the target table
+            error_on_type_mismatch (bool): specify if merge will return error if data types are mismatching :default = True
+
+        Returns:
+            TableMerger: TableMerger Object
+        """
+        invariants = self.schema().invariants
+        checker = _DeltaDataChecker(invariants)
+
+        if isinstance(source, pyarrow.RecordBatchReader):
+            schema = source.schema
+        else:
+            schema = source.schema
+
+        if isinstance(source, pyarrow.RecordBatchReader):
+            batch_iter = source
+        elif isinstance(source, pyarrow.RecordBatch):
+            batch_iter = [source]
+        elif isinstance(source, pyarrow.Table):
+            batch_iter = source.to_batches()
```

Comment:
I think for older PyArrow versions to work, this should be `to_reader()`:
```suggestion
            batch_iter = source.to_reader()
```

Reply:
As long as our minimum Pyarrowversion is 8, to_reader() is also fine. To_batches however seems to be since the very early version of Pyarrow. Not sure if there is a performance difference though.

Reply:
Yeah.